src/main/java/com/zaxxer/nuprocess/internal/BasePosixProcess.java

```diff
@@ -477,20 +477,27 @@ else if (byteBuffer != null) {
          }
       }
 
-      if (!userWantsWrite.get()) {
+      // Check whether the user has requested a write window. If so, this will clear the request and set
+      // userWantsWrite to a known-false state. Clearing the flag here ensures, if wantWrite() is called
+      // while we're in onStdinReady or, racily, after onStdinReady returns but before the result can be
+      // used to update userWantsWrite, we don't "lose" the write request
+      if (!userWantsWrite.compareAndSet(true, false)) {
          return false;
       }
 
       try {
          inBuffer.clear();
-         boolean wantMore = processHandler.onStdinReady(inBuffer);
-         userWantsWrite.set(wantMore);
+         if (processHandler.onStdinReady(inBuffer)) {
+            // If onStdinReady returns true, re-set userWantsWrite. If it returns false, we already set
+            // userWantsWrite to false before we made the callback so there's nothing to do
+            userWantsWrite.set(true);
```

Comment:
With this approach, if a user requests a write window, we clear that request and then call `onStdinReady`. If another call to `wantWrite` happens during `onStdinReady`, _regardless of what `onStdinReady` returns_, you're guaranteed another callback.

It's possible that the in-flight `onStdinReady` will actually have handled the data that the concurrent `wantWrite` wanted to send, depending on how the program is structured, which means this could result in an "extra" callback. That's unfortunate, but writing code inside a handler to prevent it, or handle it, is not that complex. The counter case, if this change isn't made, is much worse: If `onStdinReady` returns `false`, and a `wantWrite()` request happens concurrently, no `onStdinReady` callback actually happens. More importantly, it's not possible inside a handler to "fix" this race condition.

When paired with a `ReadListener` from a Java `ServletInputStream`, this can hang the process. When `ReadListener.onDataAvailable` is called, it will _never_ be called again until `isReady()` has been called and returned `false`. That means if its `wantWrite()` is dropped because it's timed just wrong with an `onStdinReady` callback (rare, certainly, but not really all that difficult to achieve in practice), `onStdinReady` is never called an `isReady` may never happen, so no subsequent `wantWrite()` will retry it.